File: plugin/lib/v4.js
// Bounded admission for host-delivered V4 logical snapshots. The official
// rc.2 catalog remains the full fixture/physical-contract validator.
import rules from './rules.js'
export const INPUT_IDENTITY = Object.freeze({target_dsh_version: '0.1.7-rc.2', input_format_version: 4, analyzer_semantics: 'v4-rc2.2'})
const surfaceTypes = new Set(['user/message', 'assistant/message', 'system/message', 'developer/message', 'tool/result'])
export function validateV4(snapshot) {
  if (snapshot?.session?.version !== 4 || !Array.isArray(snapshot.events)) throw new Error('DSH 0.1.7-rc.2 V4 snapshot required; migrate old logs upstream')
  const surface = [], markers = []
  for (const [seq, event] of snapshot.events.entries()) {
    if (event?.seq !== seq || !event.data || typeof event.data !== 'object') throw new Error('invalid V4 event sequence or data')
    if (!rules.DSH_KNOWN_RECORD_TYPES.includes(event.type)) {
      if (event.ignorable !== true) throw new Error('unknown required V4 event')
      continue
    }
    if (event.type === 'session/end-seed' && event.data.inherited === true) markers.push(seq)
    if (!surfaceTypes.has(event.type)) continue
    const m = event.type === 'user/message' ? event.data : event.data.message
    const role = event.type === 'tool/result' ? 'tool' : event.type.split('/')[0]
    if (!m || m.role !== role || typeof m.id !== 'string' || !m.id || !Array.isArray(m.content)
      || typeof m.source?.kind !== 'string' || !m.source.kind || m.source.kind === 'plugin')
      throw new Error('invalid V4 message role, identity or source')
    if (m.content.some(b => b.type === 'tool-result')) throw new Error('legacy tool-result wrapper is not V4')
    if (role === 'tool' && (typeof m.toolCallId !== 'string' || m.toolCallId !== m.source.callId || m.source.kind !== 'tool' || (m.isError !== undefined && typeof m.isError !== 'boolean')))
      throw new Error('V4 tool result call identity conflict')
    const sources = event.sourceEventSeqs
    if (sources !== undefined && (!Array.isArray(sources) || !sources.length || new Set(sources).size !== sources.length || sources.some(n => !Number.isSafeInteger(n) || n < 0 || n >= seq)))
      throw new Error('invalid V4 sourceEventSeqs')
    if (event.surfaceOp === 'append') surface.push(seq)
    else {
      const op = event.surfaceOp, start = surface.indexOf(op?.startSeq), end = surface.indexOf(op?.endSeq)
      if (op?.op !== 'replace' || start < 0 || end < start || surface.slice(start,end+1).some(n => !sources?.includes(n))) throw new Error('invalid V4 surface replacement')
      surface.splice(start,end-start+1,seq)
    }
  }
  const cut = snapshot.inheritedEventCount ?? (markers.at(-1) ?? 0)
  if (!Number.isSafeInteger(cut) || cut < 0 || cut > snapshot.events.length || (markers.at(-1) ?? 0) !== cut || (cut > 0 && !snapshot.session.isSeeded)) throw new Error('V4 inherited cut disagrees with marker')
  return cut
}

File: src/dsh_session_insights/v4.py
"""Bounded V4 admission for the read-only CLI (not an upstream migrator)."""
IDENTITY = {"target_dsh_version": "0.1.7-rc.2", "input_format_version": 4, "analyzer_semantics": "v4-rc2.2"}
SURFACE = {"user/message", "assistant/message", "system/message", "developer/message", "tool/result"}

def validate_records(records, known, *, physical=False):
    if not records or not isinstance(records[0], dict) or records[0].get("type") != "session" or records[0].get("version") != 4:
        raise ValueError("V4 header required; migrate old logs with upstream DSH tools")
    surface, markers = [], []
    for seq, event in enumerate(records[1:]):
        if not isinstance(event, dict) or event.get("seq") != seq or not isinstance(event.get("data"), dict):
            raise ValueError("invalid V4 sequence or event data")
        kind, data = event.get("type"), event["data"]
        if kind not in known:
            if event.get("ignorable") is not True:
                raise ValueError("unknown required V4 event")
            continue
        if kind == "session/end-seed" and data.get("inherited") is True:
            markers.append(seq)
        if "sourceEventSeqs" in event:
            sources = event["sourceEventSeqs"]
            if not isinstance(sources, list) or not sources:
                raise ValueError("invalid sourceEventSeqs")
            decoded, ranged = [], False
            for item in sources:
                if physical and isinstance(item, list):
                    ranged = True
                    if len(item) != 2 or any(type(n) is not int for n in item) or not 0 <= item[0] <= item[1] < seq:
                        raise ValueError("invalid sourceEventSeqs range")
                    decoded.extend(range(item[0], item[1]+1))
                elif type(item) is int and 0 <= item < seq:
                    decoded.append(item)
                else:
                    raise ValueError("invalid sourceEventSeqs reference")
            if len(set(decoded)) != len(decoded) or (ranged and decoded != sorted(decoded)):
                raise ValueError("duplicate or unordered sourceEventSeqs")
            event["sourceEventSeqs"] = decoded
        if kind not in SURFACE:
            continue
        message = data if kind == "user/message" else data.get("message")
        role = "tool" if kind == "tool/result" else kind.split("/")[0]
        if not isinstance(message, dict) or message.get("role") != role or not isinstance(message.get("id"), str) or not message["id"] or not isinstance(message.get("content"), list):
            raise ValueError("invalid V4 message identity or role")
        source = message.get("source")
        if not isinstance(source, dict) or not isinstance(source.get("kind"), str) or not source["kind"] or source["kind"] == "plugin":
            raise ValueError("invalid V4 message source")
        if any(not isinstance(b, dict) or b.get("type") == "tool-result" for b in message["content"]):
            raise ValueError("legacy tool wrapper is not V4")
        if role == "tool" and (source["kind"] != "tool" or not isinstance(message.get("toolCallId"), str) or message["toolCallId"] != source.get("callId") or ("isError" in message and type(message["isError"]) is not bool)):
            raise ValueError("V4 tool result call identity conflict")
        op = event.get("surfaceOp")
        if op == "append":
            surface.append(seq)
        else:
            if not isinstance(op, dict) or op.get("op") != "replace" or op.get("startSeq") not in surface or op.get("endSeq") not in surface:
                raise ValueError("invalid V4 surface replacement")
            start, end = surface.index(op["startSeq"]), surface.index(op["endSeq"])
            if end < start or any(n not in event.get("sourceEventSeqs", []) for n in surface[start:end+1]):
                raise ValueError("invalid V4 replacement references")
            surface[start:end+1] = [seq]
    if markers and not records[0].get("isSeeded"):
        raise ValueError("invalid V4 inherited cut")
    return markers[-1] if markers else 0
